Matrix class and pytest tests

IWorkForFood opened the pull request:
I have implemented the Matrix-class with overloaded methods and tested their functionality.

IWorkForFood committed:
Matrix class and pytest tests

egorgur commented on Matrix.py:
```diff
@@ -0,0 +1,77 @@
+class Matrix:
```
**_Write the docsrings and type annotations to class and its methods!_**

egorgur replied:
```suggestion
class Matrix:
"""
Matrix class.

Represents a mathematical matrix

"""

```

IWorkForFood replied:
I wrote these docsrings and type annotations

egorgur commented on Matrix.py:
```diff
@@ -0,0 +1,77 @@
+class Matrix:
+    def __init__(self, columns):
```
Add a docstring and type annotations

egorgur replied:
```suggestion
    def __init__(self: Self, init_data: ...):
        """Init method."""
```
the `Self` type comes from `typing` lib and you must use it to annotate the self type
this `...` **must** be the type of init_data that the new Matrix objects gets

also write the docstring to all methods and classes _but keep it simple_

IWorkForFood replied:
It is done

egorgur commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+from Matrix import Matrix
+import pytest
+
+@pytest.fixture
+def add_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    s_m = Matrix([[0, 1],[0, 2]])
+    result_m = Matrix([[0, 2],[0, 4]])
+    return [f_m, s_m, result_m]
+
+
+@pytest.fixture
+def sub_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    s_m = Matrix([[0, 1],[0, 2]])
+    result_m = Matrix([[0, 0],[0, 0]])
+    return [f_m, s_m, result_m]
+
+@pytest.fixture
+def mul_matrx_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    s_m = Matrix([[0, 1],[0, 2]])
+    result_m = Matrix([[0, 2], [0, 4]])
+    return [f_m, s_m, result_m]
+
+@pytest.fixture
+def mul_numb_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    numb = 4
+    result_m = Matrix([[0, 4],[0, 8]])
+    return [f_m, numb, result_m]
+
+@pytest.fixture
+def transpose_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    result_m = Matrix([[0, 0],[1, 2]])
+    return [f_m, result_m]
```
Why you did the tests?

egorgur replied:
Tests will be the next big task and you shoud not do it without necessary code style knowledge

IWorkForFood replied:
I ruthlessly destroyed unneccessary code

egorgur commented on Matrix.py:
```diff
@@ -0,0 +1,77 @@
+            new_matx.append([y + x for y, x in zip(f_matx, s_matx)])
+        return Matrix(new_matx)
+
+    def __iadd__(self, other):
+        return self.__add__(other)
+
+    def __sub__(self, columns):
+        columns_list = columns.matrix
+        new_matx = []
+        if len(columns_list) != len(self.matrix):
+            raise Exception("Size mismatch")
+        for f_matx, s_matx in zip(self.matrix, columns_list):
+            if len(f_matx) != len(s_matx):
+                raise Exception("Size mismatch")
+            new_matx.append([y - x for y, x in zip(f_matx, s_matx)])
+        return Matrix(new_matx)
+
+    def __isub__(self, other):
+        return self.__sub__(other)
+
+    def __mul__(self, columns):
+        if isinstance(columns, int) or isinstance(columns, float):
+            columns_list = [[columns if x == y else 0 for x in range(len(self.matrix[0]))] for y in range(len(self.matrix))]
+        else:
+            columns_list = columns.matrix
+        new_matx = []
+        if (len(self.matrix[0]) != len(columns_list)):
+            raise Exception("Size mismatch")
+
+        for fm_number in range(len(self.matrix)):
+            pos_number = []
+            for sm_number in range(len(columns_list[0])):
+                pos_number.append(sum([x * y for x, y in zip(self.matrix[fm_number], [z[sm_number] for z in columns_list])]))
+            new_matx.append(pos_number)
+            print(new_matx)
+        return Matrix(new_matx)
+
+    def __imul__(self, other):
+        return self.__mul__(other)
```
Why you overloaded more than `__add__` and `__str__` methods? Your task clearly states that you must do eather sum method or substraction method.

IWorkForFood replied:
I ruthlessly destroyed unneccessary code

egorgur commented on Matrix.py:
```diff
@@ -0,0 +1,77 @@
+        new_m = [[y[x] for y in self.matrix] for x in range(len(self.matrix))]
+        return Matrix(new_m)
+
+'''
+new_m = Matrix([[0, 1],[0, 2]])
+old_m = Matrix([[0, 1],[0, 2]])
+
+no_m = new_m + old_m
+no_p = new_m - old_m
+no_mu = new_m * old_m
+print(new_m.get_matrix())
+mul_num = new_m * 4
+tr_m = new_m.transpose()
+
+print(no_m.get_matrix(), no_p.get_matrix(), no_mu.get_matrix(), "|", mul_num.get_matrix(), "|", tr_m.get_matrix(), tr_m.size)
+'''
```
Delete commented-out code

egorgur replied:
Also, it is better to write in-dev-test in `main.py` file and not in the your-class-file.py

IWorkForFood replied:
It is done too

egorgur commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+
+@pytest.fixture
+def add_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    s_m = Matrix([[0, 1],[0, 2]])
```
Do not use short-named variables like `f_m` or `a_b`, because there are not understandable by other than you developers

egorgur commented on conftest.py:
```diff
@@ -0,0 +1,37 @@
+
+@pytest.fixture
+def sub_result():
+    f_m = Matrix([[0, 1],[0, 2]])
+    s_m = Matrix([[0, 1],[0, 2]])
```
You use those two matrices in 4 tests. Every developer **must** follow **_[DRY](https://habr.com/ru/companies/itelma/articles/546372/)_** principle

create two new fixtures like this:

```python
@pytest.fixture
def matrix_one():
    first_matrix = Matrix([[0, 1],[0, 2]])
    return first_matrix
@pytest.fixture
def matrix_two():
    second_matrix = Matrix([[0, 1],[0, 2]])
    return second_matrix 
```
**and use them in other fixtures**

egorgur commented on test_matrix.py:
```diff
@@ -0,0 +1,7 @@
+def test_matrix_operations(add_result, sub_result, mul_matrx_result, mul_numb_result, transpose_result):
+    assert (add_result[0] + add_result[1]).get_matrix() == add_result[2].get_matrix()
+    assert (sub_result[0] - sub_result[1]).get_matrix() == sub_result[2].get_matrix()
+    assert (mul_matrx_result[0] * mul_matrx_result[1]).get_matrix() == mul_matrx_result[2].get_matrix()
+    assert (mul_numb_result[0] * mul_numb_result[1]).get_matrix() == mul_numb_result[2].get_matrix()
+    assert (transpose_result[0].transpose()).get_matrix() == transpose_result[1].get_matrix()
```
You must write all those assertions as **separate tests**, _because if one assertion fails, you will never know what actually happened and what is the cause of a failure_

egorgur commented on Matrix.py:
```diff
@@ -0,0 +1,77 @@
+
+    def __mul__(self, columns):
+        if isinstance(columns, int) or isinstance(columns, float):
+            columns_list = [[columns if x == y else 0 for x in range(len(self.matrix[0]))] for y in range(len(self.matrix))]
```
Try not to write everything in one line. It is very hard to read such code

egorgur reviewed (changes requested):
Bro treneruisya

mrFroxyk reviewed (commented):
And don't forget to write the PR description :)
![image](https://github.com/user-attachments/assets/4030541b-593a-4cf7-873e-167953aad5bf)

IWorkForFood committed:
Fix issues from Matrix.py

Add docstrings, type annotations,
make code more readable, remove unnecessary methods
remove commented-out code

egorgur commented on Matrix.py:
I recommend you to rename the Matrix.py to matrix.py to stay truthful to Python naming rules conventions

IWorkForFood committed:
Rename Matrix.py to matrix.py

IWorkForFood committed:
Edit tests

IWorkForFood committed:
Merge branch 'feature/POLY-10-matrix' of https://github.com/solutions-architects/poligon-python-iworkforfood into feature/POLY-10-matrix

IWorkForFood committed:
Change several nuances concerning file naming and tests

egorgur commented on f.txt:
What is this file?

IWorkForFood replied:
Oh, forgot to delete it

egorgur commented on matrix.py:
```diff
@@ -0,0 +1,55 @@
+    """
+
+    def __init__(self: Self, columns: list[list[int | float]]) -> None:
+        """init method"""
```
Dont forget the naming conventions.
```suggestion
"""Init method."""
```
One line comment must end with a point  **.**

IWorkForFood reviewed (commented):
I followed your advice

egorgur commented on conftest.py:
```diff
@@ -0,0 +1,29 @@
+import pytest
+
+@pytest.fixture
+def matrix_one() -> Matrix:
```
Do not forget to write the docstrings
```suggestion
def matrix_one() -> Matrix:
"""Fixture to create a ``Matrix`` instance."""
```

IWorkForFood replied:
I wrote them

egorgur reviewed (changes requested):
**Great work!** and not forget to add the dockstrings ;)

IWorkForFood committed:
Change the writing of documentation

egorgur commented on f.txt:
Delete this file from the pull request

egorgur commented on test_matrix.py:
```diff
@@ -0,0 +1,9 @@
+
+def test_add_method(add_result: Matrix, matrix_one: Matrix, matrix_two: Matrix):
+    """Function for testing __add__ method."""
+    assert (matrix_one + matrix_two).get_matrix() == add_result.get_matrix()
```
I think you can overload `__eq__` method to _equal_ two matrices.
With something like this:
```python
    def __eq__(self: Self, other: Self) -> bool:
        """Equality method."""
        return self._matrix == other._matrix
```
This overload will allow this `==` behavior
```python
assert matrix_one + matrix_two == add_result
```

egorgur commented on matrix.py:
```diff
@@ -0,0 +1,55 @@
+            if len(current_column) != len(columns[0]):
+                raise Exception("Size mismatch")
+
+        self.matrix = columns
```
Make the `.matrix` private by renaming it as `._matrix`

egorgur reviewed (changes requested):
Delete the f.txt file

IWorkForFood committed:
Change [200~

"Change access mode of matrix attribute and add __eq__

egorgur reviewed (approved):
Bro, eto realno kruto!
Ty prokachalsa💪
